File: core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.deploy.history

import scala.xml.{Node, Unparsed}

import jakarta.servlet.http.HttpServletRequest

import org.apache.spark.status.api.v1.ApplicationInfo
import org.apache.spark.ui.{UIUtils, WebUIPage}
import org.apache.spark.ui.UIUtils.formatImportJavaScript

private[history] class HistoryPage(parent: HistoryServer) extends WebUIPage("") {

  def render(request: HttpServletRequest): Seq[Node] = {
    val requestedIncomplete = Option(request.getParameter("showIncomplete"))
      .getOrElse("false").toBoolean

    val displayApplications = shouldDisplayApplications(requestedIncomplete)
    val eventLogsUnderProcessCount = parent.getEventLogsUnderProcess()
    val lastUpdatedTime = parent.getLastUpdatedTime()
    val providerConfig = parent.getProviderConfig()

    val summary =
      <div class="container-fluid">
        <ul class="list-unstyled">
          {providerConfig.map { case (k, v) => <li><strong>{k}:</strong> {v}</li> }}
        </ul>
        {
          if (eventLogsUnderProcessCount > 0) {
          <p>There are {eventLogsUnderProcessCount} event log(s) currently being
            processed which may result in additional applications getting listed on this page.
            Refresh the page to view updates. </p>
          } else Seq.empty

        }
        {
          if (lastUpdatedTime > 0) {
            <p>Last updated: <span id="last-updated">{lastUpdatedTime}</span></p>
          } else Seq.empty
        }
        {
          <p>Client local time zone: <span id="time-zone"></span></p>
        }
      </div>

    val appList =
      <div class="container-fluid">
        {
          val js =
            s"""
               |${formatImportJavaScript(request, "/static/historypage.js", "setAppLimit")}
               |
               |setAppLimit(${parent.maxApplications});
               |""".stripMargin

          if (displayApplications) {
            <script src={UIUtils.prependBaseUri(
              request, "/static/dataTables.rowsGroup.js")}></script> ++
            <script type="module" src={UIUtils.prependBaseUri(
              request, "/static/historypage.js")} ></script> ++
            <script type="module">{Unparsed(js)}</script> ++ <div id="history-summary"></div>
          } else if (requestedIncomplete) {
            <h4>No incomplete applications found!</h4>
          } else if (eventLogsUnderProcessCount > 0) {
            <h4>No completed applications found!</h4>
          } else {
            <h4>No completed applications found!</h4> ++ parent.emptyListingHtml()
          }
        }
      </div>

    val pageLink =
      <div class="container-fluid">
        <a href={makePageLink(request, !requestedIncomplete)}>
          {
            if (requestedIncomplete) {
              "Back to completed applications"
            } else {
              "Show incomplete applications"
            }
          }
        </a>
        <p><a href={UIUtils.prependBaseUri(request, "/logPage/?self&logType=out")}>
          Show server log</a></p>
      </div>
    val content =
      <script type="module" src={UIUtils.prependBaseUri(
        request, "/static/historypage-common.js")}></script> ++
      <script type="module" src={UIUtils.prependBaseUri(
        request, "/static/utils.js")}></script> ++
      summary ++ appList ++ pageLink
    UIUtils.basicSparkPage(request, content, parent.title, true)
  }

  def shouldDisplayApplications(requestedIncomplete: Boolean): Boolean = {
    parent.getApplicationList().exists(isApplicationCompleted(_) != requestedIncomplete)
  }

  private def makePageLink(request: HttpServletRequest, showIncomplete: Boolean): String = {
    UIUtils.prependBaseUri(request, "/?" + "showIncomplete=" + showIncomplete)
  }

  private def isApplicationCompleted(appInfo: ApplicationInfo): Boolean = {
    appInfo.attempts.nonEmpty && appInfo.attempts.head.completed
  }
}

File: core/src/main/resources/org/apache/spark/ui/static/historypage.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

var appLimit = -1;

function setAppLimit(val) {
    appLimit = val;
}

function makeIdNumeric(id) {
  var strs = id.split("_");
  if (strs.length < 3) {
    return id;
  }
  var appSeqNum = strs[2];
  var resl = strs[0] + "_" + strs[1] + "_";
  var diff = 10 - appSeqNum.length;
  while (diff > 0) {
      resl += "0"; // padding 0 before the app sequence number to make sure it has 10 characters
      diff--;
  }
  resl += appSeqNum;
  return resl;
}

function getParameterByName(name, searchString) {
  var regex = new RegExp("[\\?&]" + name + "=([^&#]*)"),
  results = regex.exec(searchString);
  return results === null ? "" : decodeURIComponent(results[1].replace(/\+/g, " "));
}

function removeColumnByName(columns, columnName) {
  return columns.filter(function(col) {return col.name != columnName})
}

function getColumnIndex(columns, columnName) {
  for(var i = 0; i < columns.length; i++) {
    if (columns[i].name == columnName)
      return i;
  }
  return -1;
}

jQuery.extend( jQuery.fn.dataTableExt.oSort, {
    "title-numeric-pre": function ( a ) {
        var x = a.match(/title="*(-?[0-9\.]+)/)[1];
        return parseFloat( x );
    },

    "title-numeric-asc": function ( a, b ) {
        return ((a < b) ? -1 : ((a > b) ? 1 : 0));
    },

    "title-numeric-desc": function ( a, b ) {
        return ((a < b) ? 1 : ((a > b) ? -1 : 0));
    }
} );

jQuery.extend( jQuery.fn.dataTableExt.oSort, {
    "appid-numeric-pre": function ( a ) {
        var x = a.match(/title="*(-?[0-9a-zA-Z\-\_]+)/)[1];
        return makeIdNumeric(x);
    },

    "appid-numeric-asc": function ( a, b ) {
        return ((a < b) ? -1 : ((a > b) ? 1 : 0));
    },

    "appid-numeric-desc": function ( a, b ) {
        return ((a < b) ? 1 : ((a > b) ? -1 : 0));
    }
} );

jQuery.extend( jQuery.fn.dataTableExt.ofnSearch, {
    "appid-numeric": function ( a ) {
        return a.replace(/[\r\n]/g, " ").replace(/<.*?>/g, "");
    }
} );

$(document).ajaxStop($.unblockUI);
$(document).ajaxStart(function(){
    $.blockUI({ message: '<h3>Loading history summary...</h3>'});
});

$(document).ready(function() {
    $.extend( $.fn.dataTable.defaults, {
      stateSave: true,
      lengthMenu: [[20,40,60,100,-1], [20, 40, 60, 100, "All"]],
      pageLength: 20
    });

    historySummary = $("#history-summary");
    searchString = historySummary["context"]["location"]["search"];
    requestedIncomplete = getParameterByName("showIncomplete", searchString);
    requestedIncomplete = (requestedIncomplete == "true" ? true : false);

    appParams = {
      limit: appLimit,
      status: (requestedIncomplete ? "running" : "completed")
    };

    $.getJSON(uiRoot + "/api/v1/applications", appParams, function(response,status,jqXHR) {
      var array = [];
      var hasMultipleAttempts = false;
      for (i in response) {
        var app = response[i];
        if (app["attempts"][0]["completed"] == requestedIncomplete) {
          continue; // if we want to show for Incomplete, we skip the completed apps; otherwise skip incomplete ones.
        }
        var id = app["id"];
        var name = app["name"];
        if (app["attempts"].length > 1) {
            hasMultipleAttempts = true;
        }

        for (j in app["attempts"]) {
          var attempt = app["attempts"][j];
          attempt["startTime"] = formatTimeMillis(attempt["startTimeEpoch"]);
          attempt["endTime"] = formatTimeMillis(attempt["endTimeEpoch"]);
          attempt["lastUpdated"] = formatTimeMillis(attempt["lastUpdatedEpoch"]);
          attempt["log"] = uiRoot + "/api/v1/applications/" + id + "/" +
            (attempt.hasOwnProperty("attemptId") ? attempt["attemptId"] + "/" : "") + "logs";
          attempt["durationMillisec"] = attempt["duration"];
          attempt["duration"] = formatDuration(attempt["duration"]);
          var hasAttemptId = attempt.hasOwnProperty("attemptId");
          var app_clone = {"id" : id, "name" : name, "hasAttemptId" : hasAttemptId, "attempts" : [attempt]};
          array.push(app_clone);
        }
      }
      if(array.length < 20) {
        $.fn.dataTable.defaults.paging = false;
      }

      var data = {
        "uiroot": uiRoot,
        "applications": array,
        "hasMultipleAttempts": hasMultipleAttempts,
        "showCompletedColumns": !requestedIncomplete,
      }

      $.get(uiRoot + "/static/historypage-template.html", function(template) {
        var sibling = historySummary.prev();
        historySummary.detach();
        var apps = $(Mustache.render($(template).filter("#history-summary-template").html(),data));
        var attemptIdColumnName = 'attemptId';
        var startedColumnName = 'started';
        var defaultSortColumn = completedColumnName = 'completed';
        var durationColumnName = 'duration';
        var conf = {
          "columns": [
            {name: 'appId', type: "appid-numeric"},
            {name: 'appName'},
            {name: attemptIdColumnName},
            {name: startedColumnName},
            {name: completedColumnName},
            {name: durationColumnName, type: "title-numeric"},
            {name: 'user'},
            {name: 'lastUpdated'},
            {name: 'eventLog'},
          ],
          "autoWidth": false,
        };

        if (hasMultipleAttempts) {
          conf.rowsGroup = [
            'appId:name',
            'appName:name'
          ];
        } else {
          conf.columns = removeColumnByName(conf.columns, attemptIdColumnName);
        }

        var defaultSortColumn = completedColumnName;
        if (requestedIncomplete) {
          defaultSortColumn = startedColumnName;
          conf.columns = removeColumnByName(conf.columns, completedColumnName);
          conf.columns = removeColumnByName(conf.columns, durationColumnName);
        }
        conf.order = [[ getColumnIndex(conf.columns, defaultSortColumn), "desc" ]];
        conf.columnDefs = [
          {"searchable": false, "targets": [getColumnIndex(conf.columns, durationColumnName)]}
        ];
        historySummary.append(apps);
        apps.DataTable(conf);
        sibling.after(historySummary);
        $('#history-summary [data-toggle="tooltip"]').tooltip();
      });
    });
});
